Two-body-scf-guess.py

```diff
@@ -0,0 +1,112 @@
+from wrap_pyscf import *
+import numpy as np
+
+def make_monomer_dens(monomers, bs):
+    monomer_dens = []
+    for j in monomers:
+        k = density(j, bs)
+        monomer_dens.append(k)
+    return monomer_dens
+monomer_dens = make_monomer_dens(monomers, bs)
+m,m = monomer_dens[0].shape
+a = len(monomer_dens)
+
+def make_list_of_ddls(dimers, bs):
+    list_of_ddls = []
+    for d_lists in dimers:
+        list_of_ddls.append([])
+        for dimer in d_lists:
+            dens = density(dimer, bs)
+            list_of_ddls[-1].append(dens)
+    return list_of_ddls
+list_of_ddls = make_list_of_ddls(dimers, bs)
+    
+def make_z_rho(m, a):
+    z_rho = np.zeros((m*a, m*a))
+    return z_rho
+z_rho = make_z_rho(m, a)
+
+def make_ei_list(list_of_ddls):
+    ei_list = []
+    for dd_lists in list_of_ddls:
+        ei_list.append([])
+    for s in ei_list:
+        s.append([])
+        s.append([])
+    return ei_list
+ei_list = make_ei_list(list_of_ddls)
+
+def fill_ei_list(list_of_ddls, ei_list, m):
+    i_list = ei_list
+    i_list[0][0].append(np.zeros((m*2, m*2)))
+    i_list[-1][1].append(np.zeros((m*2, m*2)))
+    for s in i_list:
+        s[0].append(np.zeros((m*2, m*2)))
+        s[1].append(np.zeros((m*2, m*2)))
+    for i, dd_lists in enumerate(list_of_ddls):
+        for j, dens in enumerate(dd_lists):
+            if j < i:
+                i_list[i][0].append(dens)
+            if j >= i:
+                i_list[i][1].append(dens)
+    return i_list
+i_list = fill_ei_list(list_of_ddls, ei_list, m)
+
+def add_interacts(i_list, m):
+    overlap = []
+    for s in i_list:
+        s0 = sum(s[0])[m:m*2, m:m*2]
+        s1 = sum(s[1])[0:m, 0:m]
+        ol = np.add(s0, s1)
+        overlap.append(ol)
+    return overlap
+overlap = add_interacts(i_list, m)
+
+def compute_diags(overlap, monomer_dens, a):
+    diags = []
+    for i, ol in enumerate(overlap):
+        diag = overlap[i] - ((a-2)*monomer_dens[i])
+        diags.append(diag)
+    return diags
+diags = compute_diags(overlap, monomer_dens, a)
+
+def rho_diagonal(diags, z_rho, m):
+    for i, dgs in enumerate(diags):
+        z_rho[i*m:(i+1)*m, i*m:(i+1)*m] = diags[i]
+    rho_d = z_rho
+    return rho_d
+rho_d = rho_diagonal(diags, z_rho, m)
+
+def rho(list_of_ddls, rho_d, m):
+    for i, dd_lists in enumerate(list_of_ddls):
+        for j, dens in enumerate(dd_lists):
+            while j >= i:
+                rho_d[i*m:(i+1)*m, (j+1)*m:(j+2)*m] = list_of_ddls[i][j][0:m, m:m*2]
+                rho_d[(j+1)*m:(j+2)*m, i*m:(i+1)*m] = list_of_ddls[i][j][m:2*m, 0:m]
+                break
+    rho = rho_d
+    return rho
+rho = rho(list_of_ddls, rho_d, m)
+    
+def two_body_guess(mol, monomers, dimers, bs):
+    nitr = run_scf(mol, bs, rho)
+    return nitr
+#end of functions
```

Comment:
It's my hope that I addressed all concerns, but if I haven't I would request we settle the script over a zoom call if no one minds. This script, as (rho) is written, converges with 6 iterations. Interestingly, I tested this script with the tetramer and observed convergence after 7 cycles, a two cycle improvement from our test script. I suppose it means I made a mistake initially and corrected it with this algorithm. Sorry.

Standard (tetramer) atomic guess converges with 9 cycles!!

Reply:
It's almost there. You still have some parts of the script interspersed with the functions (e.g. lines 10-12). I also recommend you follow @jwaldrop107's advice and put your script in a block like:

```.py
if __name__ == '__main__':
```
This will allow us to directly run the script with `python3 Two-body-scf-guess.py` and to build off it in other Python scripts by using `import Two-body-scf-guess`, which will come in handy for the next step.

Now that everything is working we want to make sure we don't break it as we make changes. The easiest way to do this is with unit testing. Python provides the [unittest](https://docs.python.org/3/library/unittest.html) module for this purpose. In a separate file (say `test_two_body.py`) you should put together a basic unit test which make sure the functions in `Two-body-scf-guess.py` generate a guess density that matches an established correct answer. It probably suffices to  do this for the trimer and tetramer.

Reply:
Line 10 is where I make a non-local variable of the corresponding function. Line's 11 and 12 are needed for the other functions. Should I make a function for each of them?

Reply:
Or I suppose I could define them locally inside each function where they're needed.

Reply:
The function `two_body_guess` probably needs to be defined to take the dimers, monomers, etc and call all the other functions defined as necessary to perform the entire guess. As it is now, it's a wrapper around `run_scf` that uses the non-local `rho` for the guess. Effectively, you want `two_body_guess` to do all of the things you're doing between the function definitions.

Reply:
Or, alternatively, you want `two_body_guess` to just return the final initial guess and then you pass that to `run_scf` yourself. But that's a relatively small semantic issue.

Reply:
Thanks for clarifying that about `two_body_guess`. I'll go ahead and do that, along with your other suggestion.